functionary/ui/views/functions.py

```diff
@@ -0,0 +1,76 @@
+from django.http import HttpResponse, HttpResponseForbidden, HttpResponseRedirect
+from django.shortcuts import render
+from django.views.generic.detail import DetailView
+from django.views.generic.list import ListView
+
+from core.auth import Permission
+from core.models import Environment, Function, Task
+
+from ..forms.forms import FunctionForm
+
+
+class FunctionListView(ListView):
+    model = Function
+
+    def get_queryset(self):
+        """Filters functions not in the environment then sorts based
+        on package name, then function name."""
+
+        env_id = self.request.session["environment_id"]
+        return (
+            super()
+            .get_queryset()
+            .filter(package__environment__id=env_id)
+            .order_by("package__name", "name")
+        )
+
+
+class FunctionDetailView(DetailView):
```

Comment:
These detail views actually need the `get_queryset` stuff too.  Otherwise what happens is that if you go to a URL for a specific function, package, etc, you'll see it regardless of which environment you have active.  This is easily reproduce-able by just going to a detail page and then switching the environment.

Having said that, we clearly need a mixin or base class or something that handles all of this environment filtering for us so that it's not defined on every view.  If you want to just put a `TODO` for adding queryset filtering, that's fine.